src/transport/raw/TCP.cpp

```diff
@@ -127,21 +145,36 @@ void TCPBase::Close()
     mState = TCPState::kNotReady;
 }
 
-ActiveTCPConnectionState * TCPBase::AllocateConnection()
+ActiveTCPConnectionState * TCPBase::AllocateConnection(Inet::TCPEndPoint * endpoint, const PeerAddress & address)
 {
-    for (size_t i = 0; i < mActiveConnectionsSize; i++)
+    // If a peer initiates a connection through HandleIncomingConnection but the connection is never claimed
+    // in ProcessSingleMessage, we'll be left with a dangling ActiveTCPConnectionState which can be
+    // reclaimed.  Don't try to reclaim these connections unless we're out of space
+    for (int reclaim = 0; reclaim < 2; reclaim++)
     {
-        if (!mActiveConnections[i].InUse())
+        for (size_t i = 0; i < mActiveConnectionsSize; i++)
         {
-            return &mActiveConnections[i];
+            if (!mActiveConnections[i].InUse())
+            {
+                ActiveTCPConnectionState * activeConnection = &mActiveConnections[i];
+                VerifyOrDie(activeConnection->GetReferenceCount() == 0);
+                // Update state for the active connection
+                activeConnection->Init(endpoint, address, [this](auto & conn) { TCPDisconnect(conn, true); });
+                return activeConnection;
+            }
         }
-    }
 
+        // Out of space; reclaim connections that were never claimed by ProcessSingleMessage
+        for (size_t i = 0; i < mActiveConnectionsSize; i++)
+        {
+            ActiveTCPConnectionHolder releaseUnclaimed(&mActiveConnections[i]);
```

Comment:
So this will reclaim things that currently have a 0 refcount, basically?  Might be worth documenting that.

Reply:
Expanded on comment in line 167 in follow-up